File: #0983.minimum-cost-for-tickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {

        int size = days.size();
        vector<int> result(size);
        result[0] = std::min({costs[0], costs[1], costs[2]});
        for (int i = 1; i < size; i++) {
            int idx7 = std::upper_bound(days.begin(), days.end(), days[i]-7)-days.begin() - 1; 
            int idx30 = std::upper_bound(days.begin(), days.end(), days[i]-30)-days.begin() - 1;
            result[i] = 
                std::min({result[i-1] + costs[0],
                    (idx7 >= 0 && days[i] - days[idx7] >= 7 ? result[idx7] : 0) + costs[1],
                    (idx30 >= 0 && days[i] - days[idx30]>=30 ? result[idx30] : 0) + costs[2]
                })
            ;
        }
        return result[size - 1];
    }
};

File: #0983.minimum-cost-for-tickets.js
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function(days, costs) {
    const n = days.length;
    const dp = new Array(n).fill(0);
    dp[0] = Math.min(...costs);
    for (let i = 1; i < n; i++) {
        const oneday = dp[i-1] + costs[0];
        let sevendays = costs[1];
        let thirtydays = costs[2];
        let j = i - 1;
        while ((j >= 0) && (days[i] - days[j] < 7)) {
            j--;
        }

        if (j >= 0) {
            sevendays = dp[j] + costs[1];
        }
        j = i - 1;
        
        while ((j >= 0) && (days[i] - days[j] < 30)) {
            j--;
        }

        if (j >= 0) {
            thirtydays = dp[j] + costs[2];
        }
        dp[i] = Math.min(oneday, sevendays, thirtydays);
    }
    return dp[n - 1];
};
